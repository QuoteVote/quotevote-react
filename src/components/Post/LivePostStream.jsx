import { useEffect, useState } from 'react'
import { Grid } from '@material-ui/core'
import { useQuery, useSubscription } from '@apollo/react-hooks'
import PropTypes from 'prop-types'
import PostCard from './PostCard'
import { GET_TOP_POSTS } from '../../graphql/query'
import { NEW_POST_SUBSCRIPTION } from '../../graphql/subscription'

function LivePostStream({ paused }) {
  const { data, loading } = useQuery(GET_TOP_POSTS, {
    variables: { limit: 10, offset: 0, searchKey: '', startDateRange: '', endDateRange: '', friendsOnly: false },
    fetchPolicy: 'network-only',
  })

  const [posts, setPosts] = useState([])
  const [queued, setQueued] = useState([])

  useEffect(() => {
    if (data && data.posts) {
      setPosts(data.posts.entities)
    }
  }, [data])

  useSubscription(NEW_POST_SUBSCRIPTION, {
    onSubscriptionData: ({ subscriptionData }) => {
      const newPost = subscriptionData.data && subscriptionData.data.post
      if (!newPost) return
      if (paused) {
        setQueued((q) => [newPost, ...q])
      } else {
        setPosts((p) => [newPost, ...p])
      }
    },
  })

  useEffect(() => {
    if (!paused && queued.length) {
      setPosts((p) => [...queued, ...p])
      setQueued([])
    }
  }, [paused, queued])

  if (loading && posts.length === 0) {
    return <div>Loading...</div>
  }

  return (
    <Grid container direction="column" spacing={2}>
      {posts.map((post) => (
        <Grid item key={post._id} style={{ marginBottom: 8 }}>
          <PostCard
            {...post}
            activityType={post.activityType || 'POSTED'}
            votes={post.votes || []}
            comments={post.comments || []}
            quotes={post.quotes || []}
            messageRoom={{ messages: post.messageRoom?.messages || [] }}
            limitText={false}
          />
        </Grid>
      ))}
    </Grid>
  )
}

LivePostStream.propTypes = {
  paused: PropTypes.bool,
}

LivePostStream.defaultProps = {
  paused: false,
}

export default LivePostStream
