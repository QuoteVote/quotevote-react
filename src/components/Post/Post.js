import React, { useState } from 'react'
import {
  Card, CardActions, CardContent, CardHeader, IconButton,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import BlockIcon from '@material-ui/icons/Block'
import LinkIcon from '@material-ui/icons/Link'
import { Comment, Favorite, PersonAdd } from '@material-ui/icons'
import PropTypes from 'prop-types'
import { useDispatch } from 'react-redux'
import { useMutation } from '@apollo/react-hooks'
import { cloneDeep, findIndex } from 'lodash'
import moment from 'moment'
import ApproveButton from '../CustomButtons/ApproveButton'
import RejectButton from '../CustomButtons/RejectButton'
import VotingBoard from '../VotingComponents/VotingBoard'
import VotingPopup from '../VotingComponents/VotingPopup'
import { SET_SNACKBAR } from '../../store/ui'
import {
  ADD_COMMENT, APPROVE_POST, REJECT_POST, VOTE,
} from '../../graphql/mutations'
import { GET_POST, GET_TOP_POSTS } from '../../graphql/query'
import ApproveRejectPopover from '../ApproveRejectPopver/ApproveRejectPopover'
import AvatarDisplay from '../Avatar'

const useStyles = makeStyles(() => ({
  header2: {
    padding: 0,
    marginLeft: 10,
  },
  title: {
    color: '#00cf6e',
    marginRight: 5,
    fontFamily: 'Montserrat',
  },
  blockIcon: {
    color: 'red',
  },
  avatar: {
    marginLeft: 20,
  },
  upVote: {
    color: '#00cf6e',
  },
  downVote: {
    color: 'red',
  },
  points: {
    marginTop: 15,
    marginRight: 20,
    fontSize: 22,
    fontWeight: 'bolder',
    fontFamily: 'Montserrat',
  },
  content: {},
  expand: {
    marginLeft: 'auto',
  },
  button: {
    margin: 10,
  },
}))

function Post({ post, user }) {
  const classes = useStyles()
  const {
    title, creator, upvotes, downvotes, created, _id,
  } = post
  const { name, avatar } = creator
  const dispatch = useDispatch()
  const parsedCreated = moment(created).format('LLL')

  const [anchorEl, setAnchorEl] = useState(null)
  const [buttonType, setButtonType] = useState('approved')
  const [selectedText, setSelectedText] = useState('')
  const [addVote] = useMutation(VOTE, {
    update(
      cache,
      {
        // eslint-disable-next-line no-shadow
        data: { addVote },
      },
    ) {
      const data = cache.readQuery({
        query: GET_POST,
        variables: { postId: _id },
      })
      const clonedPost = cloneDeep(data)

      const index = findIndex(
        clonedPost.post.votedBy,
        (vote) => vote.userId === user._id,
      )
      if (index !== -1) {
        clonedPost.post.votedBy[index].type = addVote.type
        clonedPost.post.upvotes =
          addVote.type === 'up' ?
            clonedPost.post.upvotes + 1 :
            clonedPost.post.upvotes - 1

        clonedPost.post.downvotes =
          addVote.type === 'down' ?
            clonedPost.post.downvotes + 1 :
            clonedPost.post.downvotes - 1
      } else {
        clonedPost.post.votedBy.push({ type: addVote.type, userId: user._id })
        if (addVote.type === 'up') {
          clonedPost.post.upvotes++
        } else {
          clonedPost.post.downvotes++
        }
      }
      cache.writeQuery({
        query: GET_POST,
        variables: { postId: _id },
        data: { ...clonedPost },
      })
    },
    refetchQueries: [
      {
        query: GET_TOP_POSTS,
        variables: { limit: 5, offset: 0, searchKey: '' },
      },
    ],
  })
  const [addComment] = useMutation(ADD_COMMENT, {
    refetchQueries: [
      {
        query: GET_TOP_POSTS,
        variables: { limit: 5, offset: 0, searchKey: '' },
      },
      {
        query: GET_POST,
        variables: { postId: _id },
      },
    ],
  })
  const [approvePost] = useMutation(APPROVE_POST, {
    refetchQueries: [
      {
        query: GET_POST,
        variables: { postId: _id },
      },
    ],
  })
  const [rejectPost] = useMutation(REJECT_POST, {
    refetchQueries: [
      {
        query: GET_POST,
        variables: { postId: _id },
      },
    ],
  })
  const handleAddComment = async (comment, commentWithQuote = false) => {
    let startIndex
    let endIndex
    let quoteText
    if (selectedText) {
      startIndex = selectedText.startIndex
      endIndex = selectedText.endIndex
      quoteText = selectedText.text
    } else {
      startIndex = 0
      endIndex = 0
      quoteText = ''
    }

    const newComment = {
      userId: user._id,
      content: comment,
      startWordIndex: startIndex,
      endWordIndex: endIndex,
      postId: _id,
      url: post.url,
      // hashtags,
      quote: commentWithQuote ? quoteText : '',
    }

    try {
      await addComment({ variables: { comment: newComment } })
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: 'Commented Successfully',
          type: 'success',
        }),
      )
    } catch (err) {
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: `Comment Error: ${err.message}`,
          type: 'danger',
        }),
      )
    }
  }
  const handleVoting = async (type) => {
    const vote = {
      // text: selectedText.text,
      postId: post._id,
      userId: user._id,
      type,
      startWordIndex: selectedText.startIndex,
      endWordIndex: selectedText.endIndex,
    }
    try {
      await addVote({ variables: { vote } })
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: 'Voted Successfully',
          type: 'success',
        }),
      )
    } catch (err) {
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: `Vote Error: ${err.message}`,
          type: 'danger',
        }),
      )
    }
  }

  const disableApproveReject = user._id === post.userId
  const disableApprove = post.approvedBy.includes(user._id)
  const disableReject = post.rejectedBy.includes(user._id)
  const handleApprovePost = async () => {
    if (!disableApproveReject || !disableApprove) {
      try {
        dispatch(
          SET_SNACKBAR({
            open: true,
            message: 'Approved Post Successfully',
            type: 'success',
          }),
        )
        approvePost({ variables: { userId: user._id, postId: post._id } })
      } catch (err) {
        dispatch(
          SET_SNACKBAR({
            open: true,
            message: `Approve Post Error: ${err.message}`,
            type: 'danger',
          }),
        )
      }
    }
  }
  const handleRejectPost = async () => {
    if (!disableApproveReject || !disableReject) {
      try {
        rejectPost({ variables: { userId: user._id, postId: post._id } })
        dispatch(
          SET_SNACKBAR({
            open: true,
            message: 'Rejected Post Successfully',
            type: 'success',
          }),
        )
      } catch (err) {
        dispatch(
          SET_SNACKBAR({
            open: true,
            message: `Reject Post error: ${err.message}`,
            type: 'danger',
          }),
        )
      }
    }
  }
  const handlePopoverOpen = (event, type) => {
    setAnchorEl(event.currentTarget)
    setButtonType(type)
  }
  const handlePopoverClose = () => {
    setAnchorEl(null)
  }
  const pointsHeader = (
    <div className={classes.points}>
      <span className={classes.upVote}>
        +
        {upvotes}
      </span>
      <span> / </span>
      <span className={classes.downVote}>{downvotes}</span>
    </div>
  )
  const cardTitle = (
    <div>
      <span className={classes.title}>{title}</span>
      <IconButton size="small">
        <LinkIcon />
      </IconButton>
      <IconButton size="small">
        <BlockIcon className={classes.blockIcon} />
      </IconButton>
    </div>
  )
  return (
    <Card className={classes.root}>
      <CardHeader
        className={classes.header1}
        title={cardTitle}
        action={pointsHeader}
      />
      <CardHeader
        className={classes.header2}
        avatar={(
          <IconButton>
            <AvatarDisplay height={40} width={40} {...avatar} />
          </IconButton>
        )}
        title={name}
        subheader={parsedCreated}
      />
      <CardContent>
        <VotingBoard
          content={post.text}
          onSelect={setSelectedText}
          selectedText={selectedText}
          highlights
        >
          {({ text }) => (
            <VotingPopup
              onVote={handleVoting}
              onAddComment={handleAddComment}
              text={text}
              selectedText={selectedText}
              votedBy={post.votedBy}
            />
          )}
        </VotingBoard>
      </CardContent>

      <CardActions disableSpacing style={{ marginLeft: 20 }}>
        <RejectButton
          onMouseEnter={(e) => handlePopoverOpen(e, 'rejected')}
          onClick={handleRejectPost}
          style={{
            opacity: disableApproveReject || disableReject ? 0.65 : 1,
          }}
        />
        <ApproveButton
          onClick={handleApprovePost}
          style={{
            marginLeft: 10,
            opacity: disableApproveReject || disableApprove ? 0.65 : 1,
          }}
          onMouseEnter={(e) => handlePopoverOpen(e, 'approved')}
        />
        <IconButton className={classes.expand}>
          <Comment />
        </IconButton>
        <IconButton>
          <PersonAdd />
        </IconButton>
        <IconButton>
          <Favorite />
        </IconButton>
      </CardActions>

      <ApproveRejectPopover
        anchorEl={anchorEl}
        handlePopoverClose={handlePopoverClose}
        type={buttonType}
        approvedBy={post.approvedBy}
        rejectedBy={post.rejectedBy}
      />
    </Card>
  )
}

Post.propTypes = {
  post: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
}

export default Post
