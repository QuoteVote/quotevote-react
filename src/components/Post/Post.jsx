import { useState } from 'react'
import {
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import BlockIcon from '@material-ui/icons/Block'
import LinkIcon from '@material-ui/icons/Link'
import DeleteIcon from '@material-ui/icons/Delete'
import PropTypes from 'prop-types'
import { useDispatch } from 'react-redux'
import { useMutation } from '@apollo/react-hooks'
import { useHistory } from 'react-router-dom'
import { cloneDeep, findIndex, includes } from 'lodash'
import copy from 'clipboard-copy'
import moment from 'moment'
import SweetAlert from 'react-bootstrap-sweetalert'
import FollowButton from 'components/CustomButtons/FollowButton'
import VotingBoard from '../VotingComponents/VotingBoard'
import VotingPopup from '../VotingComponents/VotingPopup'
import { SET_SNACKBAR } from '../../store/ui'
import useGuestGuard from 'utils/useGuestGuard'
import RequestInviteDialog from '../RequestInviteDialog'
import {
  ADD_COMMENT,
  ADD_QUOTE,
  REPORT_POST,
  VOTE,
  APPROVE_POST,
  REJECT_POST,
  DELETE_POST
} from '../../graphql/mutations'
import {
    GET_POST,
    GET_TOP_POSTS,
    GET_USER_ACTIVITY,
} from '../../graphql/query'
import AvatarDisplay from '../Avatar'
import BookmarkIconButton from '../CustomButtons/BookmarkIconButton'
import buttonStyle from '../../assets/jss/material-dashboard-pro-react/components/buttonStyle'
import ApproveButton from '../CustomButtons/ApproveButton'
import RejectButton from '../CustomButtons/RejectButton'
import ApproveRejectPopover from '../ApproveRejectPopver/ApproveRejectPopover'
import { serializeVotedBy } from '../../utils/objectIdSerializer'

const useStyles = makeStyles(() => ({
  header2: {
    padding: 0,
    marginLeft: 10,
  },
  title: {
    color: '#00cf6e',
    marginRight: 5,
    fontFamily: 'Montserrat',
  },
  blockIcon: {
    color: 'red',
  },
  avatar: {
    marginLeft: 20,
  },
  votes: {
    color: '#00cf6e',
  },
  downVote: {
    color: 'red',
  },
  points: {
    marginTop: 15,
    marginRight: 20,
    fontSize: 22,
    fontWeight: 'bolder',
    fontFamily: 'Montserrat',
  },
  content: {},
  expand: {
    marginLeft: 'auto',
  },
  button: {
    margin: 10,
  },
  ...buttonStyle,
}))

function Post({
  post,
  user,
  postHeight,
  postActions,
}) {
  const classes = useStyles()
  const {
    title, creator, created, _id, userId,
  } = post
  const { name, avatar, username } = creator
  const { _followingId } = user
  const dispatch = useDispatch()
  const history = useHistory()
  const ensureAuth = useGuestGuard()
  const parsedCreated = moment(created).format('LLL')
  
  // State declarations
  const [selectedText, setSelectedText] = useState('')
  const [open, setOpen] = useState(false)
  const [anchorEl, setAnchorEl] = useState(null)
  const [popoverType, setPopoverType] = useState('')
  
  const isFollowing = includes(_followingId, userId)

  const handlePopoverOpen = (event, type) => {
    console.log('Popover open:', { type, anchor: event.currentTarget });
    setAnchorEl(event.currentTarget);
    setPopoverType(type);
  };

  const handlePopoverClose = () => {
    setAnchorEl(null)
    setPopoverType('')
  }

  const [addVote] = useMutation(VOTE, {
    update(
      cache,
      {
        // eslint-disable-next-line no-shadow
        data: { addVote: voteData },
      },
    ) {
      const data = cache.readQuery({
        query: GET_POST,
        variables: { postId: _id },
      })
      const clonedPost = cloneDeep(data)

      const index = findIndex(
        clonedPost.post.votedBy,
        (vote) => vote.userId === user._id,
      )
      if (index !== -1) {
        clonedPost.post.votedBy[index].type = voteData.type
        clonedPost.post.upvotes =
          voteData.type === 'up' ?
            clonedPost.post.upvotes + 1 :
            clonedPost.post.upvotes - 1

        clonedPost.post.downvotes =
          voteData.type === 'down' ?
            clonedPost.post.downvotes + 1 :
            clonedPost.post.downvotes - 1
      } else {
        clonedPost.post.votedBy.push({ type: voteData.type, userId: user._id })
        if (voteData.type === 'up') {
          clonedPost.post.upvotes++
        } else {
          clonedPost.post.downvotes++
        }
      }
      cache.writeQuery({
        query: GET_POST,
        variables: { postId: _id },
        data: { ...clonedPost },
      })
    },
    refetchQueries: [
      {
        query: GET_TOP_POSTS,
        variables: { limit: 5, offset: 0, searchKey: '' },
      },
      {
        query: GET_POST,
        variables: { postId: _id },
      },
    ],
  })
  const [addComment] = useMutation(ADD_COMMENT, {
    refetchQueries: [
      {
        query: GET_TOP_POSTS,
        variables: { limit: 5, offset: 0, searchKey: '' },
      },
      {
        query: GET_POST,
        variables: { postId: _id },
      },
    ],
  })
  const [addQuote] = useMutation(ADD_QUOTE, {
    refetchQueries: [
      {
        query: GET_TOP_POSTS,
        variables: { limit: 5, offset: 0, searchKey: '' },
      },
      {
        query: GET_POST,
        variables: { postId: _id },
      },
      {
        query: GET_USER_ACTIVITY,
        variables: {
          limit: 15,
          offset: 0,
          searchKey: '',
          activityEvent: ['POSTED', 'VOTED', 'COMMENTED', 'QUOTED', 'LIKED'],
          user_id: user._id,
          startDateRange: '',
          endDateRange: '',
        },
      },
    ],
  })

  const [reportPost] = useMutation(REPORT_POST, {
    refetchQueries: [
      {
        query: GET_TOP_POSTS,
        variables: { limit: 5, offset: 0, searchKey: '' },
      },
      {
        query: GET_POST,
        variables: { postId: _id },
      },
    ],
  })

  const [approvePost] = useMutation(APPROVE_POST, {
    refetchQueries: [
      { query: GET_POST, variables: { postId: _id } },
      { query: GET_TOP_POSTS, variables: { limit: 5, offset: 0, searchKey: '' } },
    ],
  });
  const [rejectPost] = useMutation(REJECT_POST, {
    refetchQueries: [
      { query: GET_POST, variables: { postId: _id } },
      { query: GET_TOP_POSTS, variables: { limit: 5, offset: 0, searchKey: '' } },
    ],
  });

  const userIdStr = user._id?.toString();
  const hasApproved = Array.isArray(post.approvedBy) && post.approvedBy.some(id => id?.toString() === userIdStr);
  const hasRejected = Array.isArray(post.rejectedBy) && post.rejectedBy.some(id => id?.toString() === userIdStr);

  const [removeApprove] = useMutation(APPROVE_POST, {
    variables: { postId: _id, userId: user._id },
    refetchQueries: [
      { query: GET_POST, variables: { postId: _id } },
      { query: GET_TOP_POSTS, variables: { limit: 5, offset: 0, searchKey: '' } },
    ],
  });
  const [removeReject] = useMutation(REJECT_POST, {
    variables: { postId: _id, userId: user._id },
    refetchQueries: [
      { query: GET_POST, variables: { postId: _id } },
      { query: GET_TOP_POSTS, variables: { limit: 5, offset: 0, searchKey: '' } },
    ],
  });

  const [deletePost] = useMutation(DELETE_POST, {
    refetchQueries: [
      { query: GET_TOP_POSTS, variables: { limit: 5, offset: 0, searchKey: '' } },
    ],
  });

  const handleReportPost = async () => {
    if (!ensureAuth()) return
    try {
      const res = await reportPost({ variables: { postId: _id, userId: user._id } })
      const { reportedBy } = res.data.reportPost
      const reported = reportedBy.length
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: `Post Reported. Total Reports: ${reported}`,
          type: 'success',
        }),
      )
    } catch (err) {
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: `${err.message}`,
          type: 'danger',
        }),
      )
    }
  }

  const handleAddComment = async (comment, commentWithQuote = false) => {
    if (!ensureAuth()) return
    let startIndex
    let endIndex
    let quoteText
    if (selectedText) {
      startIndex = selectedText.startIndex
      endIndex = selectedText.endIndex
      quoteText = selectedText.text
    } else {
      startIndex = 0
      endIndex = 0
      quoteText = ''
    }

    const newComment = {
      userId: user._id,
      content: comment,
      startWordIndex: startIndex,
      endWordIndex: endIndex,
      postId: _id,
      url: post.url,
      // hashtags,
      quote: commentWithQuote ? quoteText : '',
    }

    try {
      await addComment({ variables: { comment: newComment } })
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: 'Commented Successfully',
          type: 'success',
        }),
      )
    } catch (err) {
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: `Comment Error: ${err.message}`,
          type: 'danger',
        }),
      )
    }
  }
  const handleVoting = async (obj) => {
    if (!ensureAuth()) return
    const vote = {
      content: selectedText.text,
      postId: post._id,
      userId: user._id,
      type: obj.type,
      tags: obj.tags,
      startWordIndex: selectedText.startIndex,
      endWordIndex: selectedText.endIndex,
    }
    try {
      await addVote({ variables: { vote } })
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: 'Voted Successfully',
          type: 'success',
        }),
      )
    } catch (err) {
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: `Vote Error: ${err.message}`,
          type: 'danger',
        }),
      )
    }
  }
  const handleAddQuote = async () => {
    if (!ensureAuth()) return
    const quote = {
      quote: selectedText.text,
      postId: post._id,
      quoter: user._id,
      quoted: userId,
      startWordIndex: selectedText.startIndex,
      endWordIndex: selectedText.endIndex,
    }
    try {
      await addQuote({ variables: { quote } })
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: 'Quoted Successfully',
          type: 'success',
        }),
      )
    } catch (err) {
      dispatch(
        SET_SNACKBAR({
          open: true,
          message: `Quote Error: ${err.message}`,
          type: 'danger',
        }),
      )
    }
  }

  const handleRedirectToProfile = (profileUsername) => {
    history.push(`/Profile/${profileUsername}`)
  }

  const pointsHeader = (
    <div className={classes.points}>
      <span className={classes.votes}>
        {postActions ? postActions.length : '0'}
      </span>
    </div>
  )

  const copyToClipBoard = async () => {
    const baseUrl = window.location.origin
    await copy(`${baseUrl}${history.location.pathname}`)
    setOpen(true)
    // navigator.clipboard.writeText(`www.quote.vote${history.location.pathname}`)
  }

  const hideAlert = () => {
    setOpen(false)
  }

  const cardTitle = (
    <div>
      <span className={classes.title}>{title}</span>
      <IconButton size="small" id="copyBtn" onClick={copyToClipBoard}>
        <LinkIcon />
      </IconButton>
      <IconButton size="small" onClick={handleReportPost}>
        <BlockIcon className={classes.blockIcon} />
      </IconButton>
    </div>
  )

  const handleApprovePost = async () => {
    if (!ensureAuth()) return
    if (hasApproved) {
      // Remove approval (toggle off)
      try {
        await removeApprove({ variables: { postId: _id, userId: user._id, remove: true } });
        dispatch(
          SET_SNACKBAR({ open: true, message: 'Approval removed', type: 'success' })
        );
      } catch (err) {
        dispatch(
          SET_SNACKBAR({ open: true, message: `Approve Error: ${err.message}`, type: 'danger' })
        );
      }
    } else {
      // Approve (and override reject if needed)
      try {
        await approvePost({ variables: { postId: _id, userId: user._id } });
        dispatch(
          SET_SNACKBAR({ open: true, message: 'Post Approved', type: 'success' })
        );
      } catch (err) {
        dispatch(
          SET_SNACKBAR({ open: true, message: `Approve Error: ${err.message}`, type: 'danger' })
        );
      }
    }
  };

  const handleRejectPost = async () => {
    if (!ensureAuth()) return
    if (hasRejected) {
      // Remove rejection (toggle off)
      try {
        await removeReject({ variables: { postId: _id, userId: user._id, remove: true } });
        dispatch(
          SET_SNACKBAR({ open: true, message: 'Rejection removed', type: 'success' })
        );
      } catch (err) {
        dispatch(
          SET_SNACKBAR({ open: true, message: `Reject Error: ${err.message}`, type: 'danger' })
        );
      }
    } else {
      // Reject (and override approve if needed)
      try {
        await rejectPost({ variables: { postId: _id, userId: user._id } });
        dispatch(
          SET_SNACKBAR({ open: true, message: 'Post Rejected', type: 'success' })
        );
      } catch (err) {
        dispatch(
          SET_SNACKBAR({ open: true, message: `Reject Error: ${err.message}`, type: 'danger' })
        );
      }
    }
  };

  const handleDelete = async () => {
    try {
      await deletePost({ variables: { postId: _id } });
      dispatch(
        SET_SNACKBAR({ open: true, message: 'Post deleted', type: 'success' }),
      );
      history.push('/home');
    } catch (err) {
      dispatch(
        SET_SNACKBAR({ open: true, message: `Delete Error: ${err.message}`, type: 'danger' }),
      );
    }
  };

  return (
    <>
      <Card
        style={{
          height: postHeight >= 742 ? '83vh' : 'auto',
          overflow: 'auto',
        }}
      >
        <CardHeader
          className={classes.header1}
          title={cardTitle}
          action={pointsHeader}
        />
        <CardHeader
          className={classes.header2}
          avatar={(
            <IconButton
              size="small"
              onClick={() => handleRedirectToProfile(creator.username)}
            >
              <AvatarDisplay height={40} width={40} {...avatar} />
            </IconButton>
          )}
          title={name}
          subheader={parsedCreated}
        />
        <CardContent>
          <VotingBoard
            content={post.text}
            onSelect={setSelectedText}
            selectedText={selectedText}
            highlights
          >
            {({ text }) => (
              <VotingPopup
                onVote={handleVoting}
                onAddComment={handleAddComment}
                onAddQuote={handleAddQuote}
                text={text}
                selectedText={selectedText}
                votedBy={serializeVotedBy(post.votedBy)}
              />
            )}
          </VotingBoard>
        </CardContent>

        <CardActions disableSpacing style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginLeft: 20 }}>
          <div style={{ display: 'flex', gap: 8 }}>
            <RejectButton
              onMouseOver={(e) => handlePopoverOpen(e, 'rejected')}
              onClick={handleRejectPost}
              selected={hasRejected}
            />
            <ApproveButton
              onMouseOver={(e) => handlePopoverOpen(e, 'approved')}
              onClick={handleApprovePost}
              selected={hasApproved}
            />
            <ApproveRejectPopover
              anchorEl={anchorEl}
              handlePopoverClose={handlePopoverClose}
              type={popoverType}
              rejectedBy={post.rejectedBy}
            />
          </div>
          <div style={{ display: 'flex', gap: 8 }}>
            <FollowButton
              isFollowing={isFollowing}
              profileUserId={userId}
              username={username}
              showIcon
            />
            <BookmarkIconButton post={post} user={user} />
            {/* Add chat, person, and heart icons here as needed */}
          </div>
        </CardActions>
        {open && (
          <SweetAlert
            confirmBtnCssClass={`${classes.button} ${classes.success}`}
            success
            onConfirm={hideAlert}
            onCancel={hideAlert}
            title="Post URL copied!"
            timeout={1000}
          />
          <BookmarkIconButton post={post} user={user} />
          {(user._id === userId || user.admin) && (
            <IconButton onClick={handleDelete} size="small">
              <DeleteIcon />
            </IconButton>
          )}
          {/* Add chat, person, and heart icons here as needed */}
        </div>
      </CardActions>
      {open && (
        <SweetAlert
          confirmBtnCssClass={`${classes.button} ${classes.success}`}
          success
          onConfirm={hideAlert}
          onCancel={hideAlert}
          title="Post URL copied!"
          timeout={1000}
        />
      )}
    </Card>
    <RequestInviteDialog open={openInvite} onClose={() => setOpenInvite(false)} />
  )
}

Post.propTypes = {
  postActions: PropTypes.array,
  post: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  postHeight: PropTypes.number,
}

export default Post
