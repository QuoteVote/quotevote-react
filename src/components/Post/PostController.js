import React, { useEffect, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useParams } from 'react-router'
import { useQuery } from '@apollo/react-hooks'
import { GET_POST } from 'graphql/query'
import PostPage from 'views/PostsPage/PostPage'
import { SET_SELECTED_PAGE } from 'store/ui'

function PostController() {
  //  Set state for events and use viewModel props for redux/apollo?
  const conditions = ['POSTED', 'VOTED', 'COMMENTED', 'QUOTED']
  const [selectedEvent, setSelectedEvent] = useState(conditions)
  const [selectAll, setSelectAll] = useState('ALL')
  const [offset, setOffset] = useState(1)
  const limit = 5
  const { id } = useParams()
  console.log("POSTCONTROLLER: ")
  console.log(id)
  const [userInfo, setUserInfo] = useState({})
  const loggedInUser = useSelector((state) => state.user.data)

  const dispatch = useDispatch()
  const filterState = useSelector((state) => state.filter)
  const handleActivityEvent = (event, newActivityEvent) => {
    if (!newActivityEvent.length) {
      setSelectAll(['ALL'])
      setSelectedEvent(conditions)
    } else {
      const isAllToggled = newActivityEvent.length === 4
      setSelectAll(isAllToggled ? ['ALL'] : [])
      setSelectedEvent(newActivityEvent)
    }
  }

  const handleSelectAll = (event, newSelectAll) => {
    if (newSelectAll.length) {
      setSelectedEvent(conditions)
    }
    setSelectAll(newSelectAll)
  }

  const { data: postData } = useQuery(GET_POST, {
    variables: { postId: id },
  })

  useEffect(() => {
    dispatch(SET_SELECTED_PAGE(null))
  }, [dispatch])

  useEffect(() => {
    if (postData) {
      setUserInfo(postData)
    }
  }, [postData])

  return (
    <PostPage
      handleActivityEvent={handleActivityEvent}
      handleSelectAll={handleSelectAll}
      selectAll={selectAll}
      filterState={filterState}
      dispatch={dispatch}
      setOffset={setOffset}
      profileUser={userInfo}
      limit={limit}
      offset={offset}
      selectedEvent={selectedEvent}
    />
  )
}

export default PostController
