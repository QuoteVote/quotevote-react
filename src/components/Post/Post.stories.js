// Important stuff, must always be imported on a storybook file
import React from 'react'
import { withKnobs } from '@storybook/addon-knobs/react'
import { withA11y } from '@storybook/addon-a11y'

// Apollo Imports not needed for story, just for calling data
// If you want to apply theme
// The component
import Post from './Post'
import PostSkeleton from './PostSkeleton'

// Story config
export default {
  title: 'Post',
  component: Post,
  decorators: [withKnobs, withA11y],
}

const post = {
  creator: {
    name: 'John Doe',
    avatar: 'J',
  },
  created: '11/06/2018 11:00 AM',
  title: 'Title of a post',
  text: 'What is Lorem Ipsum? Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.',
  upvotes: 100,
  downvotes: 200,
  approvedBy: [],
  rejectedBy: [],
}

const user = {
  name: 'John Doe',
  avatar: 'J',
}

export const Base = () => <Post post={post} user={user} />
export const Loading = () => <PostSkeleton />

Base.story = {
  parameters: {
    jest: ['Post.test.js'],
  },
}
