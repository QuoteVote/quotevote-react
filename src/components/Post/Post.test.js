import React from 'react'
import { render } from '@testing-library/react'

// Component being tested
import Post from './Post'
import withTestWrapper from '../../hoc/withTestWrapper'

const SubmitPostWrapper = withTestWrapper(Post)

describe('Post test -', () => {
  it('renders correctly', () => {
    const { container } = render(
      <SubmitPostWrapper />,
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
