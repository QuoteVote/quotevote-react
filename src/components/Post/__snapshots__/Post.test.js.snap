// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post test - renders correctly 1`] = `null`;
